### PR TITLE
Add/remove workflow step [PREMIUM-194]

### PR DESCRIPTION
**Pull request by JanJakes**
Sorry for another huge PR. This got much bigger than I thought. I tried hard to make the commits nice so going commit-by-commit should make sense. Premium PR included as well.

**Testing:** With Premium activated and a valid Premium key, it should be possible to add/remove steps and triggers within a workflow and save the workflow.

[PREMIUM-194]
[MAILPOET-4585]

[PREMIUM-194]: https://mailpoet.atlassian.net/browse/PREMIUM-194?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[MAILPOET-4585]: https://mailpoet.atlassian.net/browse/MAILPOET-4585?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: mailpoet/lib/Automation/Engine/API/API.php [hunk 1]
```diff
@@ -58,24 +58,29 @@ private function registerRoute(string $route, string $endpointClass, string $met
...
-        $endpoint = $this->endpointContainer->get($endpointClass);
-        return $endpoint->checkPermissions();
-      },
-      'args' => $schema,
-    ]);
+          return $endpoint->checkPermissions();
+        },
+        'args' => $schema,
+      ],
+      true // allow overriding
```
Comment (websupporter) [inline mailpoet/lib/Automation/Engine/API/API.php]: It seems, allowing to overwrite in this way removes the free plugin `GET workflows` in favor of the the premium `POST workflows` route.

At least in my tests, setting the allow overwrite value to `false` recovers the `GET` endpoint

image

**Comment by websupporter** [comment]
General question:
> it should be possible to add/remove [...] triggers

How can I do that? I miss somehow the obvious, but a completely empty workflow seems not to provide me a possibility and I do not see one in a workflow populated with steps
image
image

**Comment by websupporter** [comment]
I wonder, now that there is a empty trigger necessary as a minimum for a workflow; in theory you can send `step:[]` to the API and a workflow is created without an empty trigger and basically useless. I think we should either throw an error from the API for such a case or fill an empty trigger as a default in case of `[]`

image

File: mailpoet/assets/js/src/automation/editor/components/inserter-popover/index.tsx [hunk 2]
```diff
@@ -1,28 +1,66 @@
...
+import { useDispatch, useSelect } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
+import { Hooks } from 'wp-js-hooks';
+import { PremiumModal } from 'common/premium_modal';
 import { Inserter } from '../inserter';
-import { store } from '../../store';
+import { Item } from '../inserter/item';
+import { storeName } from '../../store';
 
 export function InserterPopover(): JSX.Element | null {
```
Comment (websupporter) [inline mailpoet/assets/js/src/automation/editor/components/inserter-popover/index.tsx:11]: `JSX.Element | null`

out of curiosity: Isn't `null` a valid `JSX.Element`?

File: mailpoet/lib/Automation/Engine/API/API.php (hunk 1 above)
Comment (JanJakes) [reply mailpoet/lib/Automation/Engine/API/API.php]: Yep, good catch! I need to overwrite the PUT route but not the others, damn. I guess I can use some filter to replace it instead. Checking...

Comment (JanJakes) [reply mailpoet/lib/Automation/Engine/API/API.php]: Ahh, the fix is actually pretty simple! I'll drop this commit and make the premium route registrations run before the free ones!

File: mailpoet/assets/js/src/automation/editor/components/inserter-popover/index.tsx (hunk 2 above)
Comment (JanJakes) [reply mailpoet/assets/js/src/automation/editor/components/inserter-popover/index.tsx:11]: It's [a bit compicated](https://stackoverflow.com/a/58123882/1236795), but no, `null` is not included in `JSX.Element`, it's only included in `React.ReactNode` but JSX element [seems to be the best consensus](https://github.com/typescript-cheatsheets/react/issues/129) for typing component return types.

**Comment by JanJakes** [comment]
> I wonder, now that there is a empty trigger necessary as a minimum for a workflow; in theory you can send step:[] to the API and a workflow is created without an empty trigger and basically useless. I think we should either throw an error from the API for such a case or fill an empty trigger as a default in case of []

Good point! There are more things we don't validate yet, and I want to prepare a ticket for workflow validation, so I would like to include that. I'm not yet fully sure if we'll really need the empty triggers when we improve the workflow format to support multiple next steps (conditions, A/B test, etc.) but I want to look at it next week so I'll keep this in mind.

File: mailpoet/lib/Automation/Engine/API/API.php (hunk 1 above)
Comment (websupporter) [reply mailpoet/lib/Automation/Engine/API/API.php]: Thanks! Seems to do the trick 
image

**Review by websupporter — DISMISSED** [review]
Thanks for the PRs. Really nice work. LGTM

**Review by veljkho — APPROVED** [review]

**Pull request merged**